### ISSUE TITLE
Don't share `public` folder between dev and build

### ISSUE THREAD

**Issue by borekb** [gatsbyjs/gatsby#4512]
I've been working with Gatsby for the past few days and have already been bitten by the fact that the `/public` folder is shared between the dev server and the eventual build. Maybe as a workaround I could use something like `gatsby build --output-dir dist` as proposed in #1878 but I still think it's generally a bad idea to let dev server influence the build and vice versa (it happened to me).

**Comment by KyleAMathews** [comment]
What problems specifically did this cause you?

**Comment by borekb** [comment]
It was around loading external `<script>`s (think jQuery) which is generally tricky in React and may behave differently between dev server and the generated site. In my specific case, the scripts were not loaded on the dev server, then I ran the build in a separate console window and suddenly, the dev started working.

Maybe riskier is the other way around: if I'm not aware of this shared folder dependency, I would expect that when I run `gatsby build` and then deploy the `public` folder, it would surely be my built site. This is not guaranteed if I forgot that a dev server is running in another console.

You could say that my workflow was problematic or that I didn't understand enough of how `<script>` works in React, and you'd be right. Still, I think that production build simply should not be influenced by a dev server.

**Comment by KyleAMathews** [comment]
Hmmm I'm still not sure what happened.

Generally though, the workflow we support is importing the jquery (or other packages) from NPM and letting webpack bundle your JavaScript together. You should rarely if ever be writing `<script>` tags.

So:

```
npm install --save jquery

# In a react component where you use jquery
import jquery from "jquery"
```

If you can figure out how the dev server and production builds interacted — it'd be great to fix that.

**Comment by borekb** [comment]
Thanks. Maybe a question: when I do `gatsby develop` and there was previously something generated in `public` by `gatsby build`, will anything be overwritten? Wait, I won't be lazy and try..

**Comment by borekb** [comment]
Yeah, the differences are vast. So the problematic scenario from above can easily happen.

![image](https://user-images.githubusercontent.com/101152/37415808-36228ae4-27a4-11e8-8157-33853f8637bf.png)
